modules/040-control-plane-manager/images/control-plane-manager/controller/kubeconfig.go

```diff
@@ -65,73 +65,77 @@ func renewKubeconfig(componentName string) error {
 	path := filepath.Join(kubernetesConfigPath, componentName+".conf")
 	log.Infof("generate or renew %s kubeconfig", path)
 	if _, err := os.Stat(path); err == nil && config.ConfigurationChecksum != config.LastAppliedConfigurationChecksum {
-		var remove bool
 		tmpPath := filepath.Join(config.TmpPath, path)
 		log.Infof("configuration has changed since last kubeconfig generation (last applied checksum %s, configuration checksum %s), verifying kubeconfig", config.LastAppliedConfigurationChecksum, config.ConfigurationChecksum)
 		if err := prepareKubeconfig(componentName, true); err != nil {
 			return err
 		}
 
-		currentKubeconfig, err := loadKubeconfig(path)
+		err, remove := validateKubeconfig(path, tmpPath)
 		if err != nil {
-			return err
+			log.Error(err)
 		}
-		tmpKubeconfig, err := loadKubeconfig(tmpPath)
-		if err != nil {
-			return err
+		if remove {
+			removeFile(path)
 		}
+	}
 
-		if len(currentKubeconfig.Clusters) == 0 {
-			return fmt.Errorf("clusters field of kubeconfig %s is empty", path)
-		}
+	if _, err := os.Stat(path); err == nil {
+		return nil
+	}
 
-		if len(tmpKubeconfig.Clusters) == 0 {
-			return fmt.Errorf("clusters field of kubeconfig %s is empty", tmpPath)
-		}
+	// regenerate kubeconfig
+	log.Infof("generate new kubeconfig %s", path)
+	return prepareKubeconfig(componentName, false)
+}
 
-		if currentKubeconfig.Clusters[0].Cluster.Server != tmpKubeconfig.Clusters[0].Cluster.Server {
-			log.Infof("kubeconfig %s address field changed", path)
-			remove = true
-		}
+func validateKubeconfig(path string, tmpPath string) (error, bool) {
```

Comment:
it's wrong way - you should return error and check error type. You do not need bool parameter here.

Reply:
https://github.com/deckhouse/deckhouse/pull/9445/commits/db5f1c108673fb1e77e5159de99597c41ba28362
please review